problemtools/verifyproblem.py

```diff
@@ -1239,37 +1252,72 @@ def __init__(self, problem):
     def __str__(self):
         return 'submissions'
 
-    def check_submission(self, sub, args, expected_verdict, timelim_low, timelim_high):
+    def check_submission(self, sub, args, expected_verdict, timelim, timelim_low, timelim_high):
         desc = '%s submission %s' % (expected_verdict, sub)
-        (result1, result2) = self._problem.testdata.run_submission(sub, args, timelim_low, timelim_high)
-
-        if result1.verdict != result2.verdict:
-            r1, r2 = result1.verdict, result2.verdict
-            self.warning('%s sensitive to time limit: limit of %s secs -> %s, limit of %s secs -> %s' % (desc, timelim_low, r1, timelim_high, r2))
-
-        if result1.verdict == expected_verdict:
+        partial = False
+        if expected_verdict == 'PAC':
+            # For partially accepted solutions, use the low timelim instead of the real one,
+            # to make sure we have margin in both directions.
+            expected_verdict = 'AC'
+            partial = True
+            timelim = timelim_low
+
+        (result1, result2) = self._problem.testdata.run_submission(sub, args, timelim, timelim_high)
+
+        if result1.verdict == 'AC' and expected_verdict == 'AC' and not partial and result1.sample_failures:
+            res = result1.sample_failures[0]
+            self.warning('%s got %s on sample: %s' % (desc, res.verdict, res))
+
+        if result1.verdict != result2.verdict or result1.score != result2.score:
+            r1, r2 = result1, result2 if result1.verdict == result2.verdict else result1.verdict, result2.verdict
+            self.warning('%s sensitive to time limit: limit of %s secs -> %s, limit of %s secs -> %s' % (desc, timelim, r1, timelim_high, r2))
+
+        if partial and self.fully_accepted(result1):
+            self.warning('%s got %s' % (desc, result1))
+        elif result1.verdict == expected_verdict:
             self.msg('   %s OK: %s' % (desc, result1))
-        elif result2.verdict == expected_verdict:
+            if (expected_verdict == 'AC' and not partial
+                    and not self.fully_accepted(result1)
+                    and self.full_score_attainable()):
+                # For some heuristic problems, this is expected. Thus, only warn.
+                self.warning('%s did not attain full score (consider moving it to partially_accepted)' % desc)
+        elif result2.verdict == expected_verdict and not (partial and self.fully_accepted(result2)):
             self.msg('   %s OK with extra time: %s' % (desc, result2))
         else:
             self.error('%s got %s' % (desc, result1), result2.additional_info)
 
         return result1
 
+    def full_score_attainable(self):
```

Comment:
Nitpick: I would call this `full_score_finite` instead, because it is more immediately obvious what it means and more accurate (the name `attainable` hints that it somehow tries to solve the undecidable problem of figuring out which scores the grader might produce).